README.md
# ARRISPWOD - Arris Password Of Day

[![Build Status](https://travis-ci.org/jesusgoku/arrispwod.svg?branch=master)](https://travis-ci.org/jesusgoku/arrispwod)
[![codecov](https://codecov.io/gh/jesusgoku/arrispwod/branch/master/graph/badge.svg)](https://codecov.io/gh/jesusgoku/arrispwod)

Javascript library for generate password of the day for Arris routers.

## Install

```bash
npm install arrispwod

# Or with yarn
yarn add arrispwod
```

## Usage

### NodeJS

```javascript
// CommonJS
const genPassOfDay = require('arrispwod');

// Or with ESNext
import genPassOfDay from 'arrispwod';


const passOfDay = getPassOfDay(new Date());
console.log(passOfDay);
```

### Browser

Use UMD package from: [https://unpkg.com/arrispwod](https://unpkg.com/arrispwod)

## API

See API [here](docs/api.md)

## Related

- [arrispwod-cli - Utility for CLI](https://www.npmjs.com/package/arrispwod-cli)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [1.1.0](https://github.com/jesusgoku/arrispwod/compare/v1.0.2...v1.1.0) (2020-04-11)


### Features

* add and setting commitlint and husky ([c09a061](https://github.com/jesusgoku/arrispwod/commit/c09a061a08183194084c2bc67027d58b0fba98fc))
* add and setting standard-version ([22c14cf](https://github.com/jesusgoku/arrispwod/commit/22c14cfa6ec8e42c1a739cd7cf2a7caa761be8b3))
* add API documentation ([3a3a1f8](https://github.com/jesusgoku/arrispwod/commit/3a3a1f810a0cb48cf747d8f802a16804509c6466))
* build with rollup ([66da012](https://github.com/jesusgoku/arrispwod/commit/66da012e345f1582d9054d77e78d263ab7146f90))


test/index.js
const expect = require('chai').expect;
const genPassOfDay = require('../src');

describe('Testing genPassOfDay function', () => {

  it('With default seed', () => {
    expect(genPassOfDay(new Date('2018-04-22 00:00:00')))
      .to
      .equal('ZOU3MWLZKX')
    ;
  });

  it('With vtr2014 seed', () => {
    expect(genPassOfDay(
      new Date('2018-04-22 00:00:00'),
      'vtr2014',
    ))
      .to
      .equal('Y20E72FQ9I')
    ;
  });

  it('With one month with vtr2014 seed', () => {
    const testCases = {
      '2018-04-01': 'ZVUUQ7B467',
      '2018-04-02': 'X84F1DZ4N6',
      '2018-04-03': 'Q7S3502K5I',
      '2018-04-04': 'ZA8ICZEDFN',
      '2018-04-05': '7C0EJIY7AU',
      '2018-04-06': 'OO8IHECY7P',
      '2018-04-07': 'I7SEDJ05ZC',
      '2018-04-08': '43RN6SAEXU',
      '2018-04-09': 'W1T3B2G82M',
      '2018-04-10': 'W1TUZAQWIY',
      '2018-04-11': 'W8DFFJZCWX',
      '2018-04-12': 'FQ3U17NECW',
      '2018-04-13': 'BD5F08ZQYV',
      '2018-04-14': '1W53ZLMBUE',
      '2018-04-15': '4X1F0XB40T',
      '2018-04-16': 'A5VU57EXOS',
      '2018-04-17': 'VI0NY4DKUT',
      '2018-04-18': '5SS3BGGHAS',
      '2018-04-19': 'TF2NGYA9D3',
      '2018-04-20': '7Q73B1L9OH',
      '2018-04-21': 'R94NBG2YEE',
      '2018-04-22': 'Y20E72FQ9I',
      '2018-04-23': 'N22IC0VZ66',
      '2018-04-24': 'WUGFZ3N79K',
      '2018-04-25': 'AB9UEFNWOJ',
      '2018-04-26': 'EDDF9SH4BI',
      '2018-04-27': 'F0UUGTFBOH',
      '2018-04-28': 'FUQIICQI9B',
      '2018-04-29': 'ZPUUI7B40F',
      '2018-04-30': 'X84F17T4FE',
    };

    Object
      .keys(testCases)
      .forEach(date => {
        expect(genPassOfDay(
          new Date(`${date} 12:00:00`),
          'vtr2014',
        ))
          .to
          .equal(testCases[date])
        ;
      })
    ;
  });

  it('Throw error when not pass date instance', () => {
    expect(() => genPassOfDay('2018-04-01 00:00:00'))
      .to
      .throw('Date is not a Date instance')
    ;
  });

  it('Throw error when seed is not string', () => {
    expect(() => genPassOfDay(new Date('2018-04-01 00:00:00'), 2014))
      .to
      .throw('Seed is not a String instance')
    ;
  });

  it('Throw error with empty seed', () => {
    expect(() => genPassOfDay(new Date('2018-04-01 00:00:00'), ''))
      .to
      .throw('Seed min length: 1')
    ;
  });
});


src/index.js
const DEFAULT_SEED = 'MPSJKMDHAI';

const TABLE1 = [
  [15, 15, 24, 20, 24],  // Monday
  [13, 14, 27, 32, 10],  // Tuesday
  [29, 14, 32, 29, 24],  // Wendesday
  [23, 32, 24, 29, 29],  // Thursday
  [14, 29, 10, 21, 29],  // Friday
  [34, 27, 16, 23, 30],  // Saturday
  [14, 22, 24, 17, 13],  // Sunday
];

const TABLE2 = [
  [0, 1, 2, 9, 3, 4, 5, 6, 7, 8],
  [1, 4, 3, 9, 0, 7, 8, 2, 5, 6],
  [7, 2, 8, 9, 4, 1, 6, 0, 3, 5],
  [6, 3, 5, 9, 1, 8, 2, 7, 4, 0],
  [4, 7, 0, 9, 5, 2, 3, 1, 8, 6],
  [5, 6, 1, 9, 8, 0, 4, 3, 2, 7],
];

const ALPHANUM = [
  '0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D',
  'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R',
  'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z'
];

/**
 * Generate Pass of the day for Arris Router
 *
 * @param {Date} [day] - Date for calculate password.
 * @param {string} [seed] - Seed for generate password. Min length: 1
 *
 * @returns {string} - password of the day
 **/
function genPassOfDay(day = new Date(), seed = DEFAULT_SEED) {
  if (!(day instanceof Date)) {
    throw new TypeError('Date is not a Date instance');
  }

  if (typeof seed !== 'string') {
    throw new TypeError('Seed is not a String instance');
  }

  if (seed.length < 1) {
    throw new Error('Seed min length: 1');
  }

  const seedRepeated = seed.repeat(10);
  const year = day.getFullYear() % 100; // Two last digits
  const month = day.getMonth() + 1; // January === 1
  const dayOfMonth = day.getDate();
  const dayOfWeek = day.getDay() === 0 ? 6 : day.getDay() - 1; // Monday === 1

  const l1 = TABLE1[dayOfWeek].slice(0);
  l1.push(dayOfMonth);
  l1.push(year + month - dayOfMonth < 0
    ? (year + month - dayOfMonth + 36) % 36
    : (year + month - dayOfMonth) % 36);
  l1.push((3 + (year + month) % 12) * dayOfMonth % 37 % 36);

  const l2 = [...Array(8).keys()].map((_, i) => seedRepeated.charCodeAt(i) % 36);

  const l3 = l1.map((f, i) => (f + l2[i]) % 36);
  const l38 = l3.reduce((t, i) => t + i, 0) % 36;
  const num8 = l38 % 6;
  l3.push(l38);
  l3.push(Math.round(Math.pow(num8, 2)));

  const l4 = TABLE2[num8].map(i => l3[i]);

  const l5 = l4.map((v, i) => (seedRepeated.charCodeAt(i) + v) % 36);

  return l5.map(i => ALPHANUM[i]).join('');
}

module.exports = genPassOfDay;
